Initializing global arrays not working

<!-- Thanks for reporting a bug for this project. READ THIS FIRST:

- Provide as many details as possible. Simply saying "X gives bug" or "X gives error" is not enough!
- Paste logs, configuration sample and code into the backticks (```).
- Read through the template carefully and fill out all missing details.
- Please also search for similar issues in this issue tracker first and read through the ESPHome FAQ.

DO NOT DELETE ANY TEXT from this template! Otherwise the issue may be closed without a comment.
-->

**Operating environment/Installation (Hass.io/Docker/pip/etc.):**
<!--
Please provide details about your environment below this line. -->
PIP

**ESP (ESP32/ESP8266, Board/Sonoff):**
<!--
Please provide details about which ESP you're using below.
-->
Sonoff

**Affected component:**
<!--
Please add the link to the documentation at https://esphome.io/index.html of the component in question.
-->
https://esphome.io/guides/automations.html?highlight=globals

**Description of problem:**
I can't compile while trying to initialize global array of integer (int[4])

**Problem-relevant YAML-configuration entries:**
```yaml
globals:
  - id: gpio0_relays
    type: int[4]
    restore_value: no
    initial_value: '{1,2,3,4}'
  - id: gpio14_relays
    type: int[4]
    restore_value: no
    initial_value: '{1,2,3,4}'
```

**Traceback (if applicable):**
<!--
Please copy the traceback here if compilation is failing. If possible, also connect to the ESP over USB and copy its logs into the backticks.
-->
```
PS F:\Home Automation\@Projects\ESPHome> py -2 -m esphome test_switch.yaml compile
[32mINFO Reading configuration...[0m
[32mINFO Generating C++ source...[0m
[32mINFO Compiling app...[0m
[32mINFO Running:  platformio run -d esphome_test[0m
Processing esphome_test (platform: espressif8266@1.8.0; framework: arduino; board: esp01_1m)
-----------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif8266/esp01_1m.html
PLATFORM: Espressif 8266 > Espressif Generic ESP8266 ESP-01 1M
HARDWARE: ESP8266 80MHz 80KB RAM (1MB Flash)
Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF MODES: FINDER(chain) COMPATIBILITY(soft)
Collected 29 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <ESP Async WebServer> 1.1.1
|   |-- <ESPAsyncTCP> 1.2.0
|   |   |-- <ESP8266WiFi> 1.0
|   |-- <ESP8266WiFi> 1.0
|   |-- <Hash> 1.0
|   |-- <ArduinoJson-esphomelib> 5.13.3
|-- <ESP8266WiFi> 1.0
|-- <ESP8266mDNS>
|   |-- <ESP8266WiFi> 1.0
|-- <ESPAsyncTCP> 1.1.3
|   |-- <ESP8266WiFi> 1.0
|-- <Hash> 1.0
|-- <esphome-core> 1.11.2
|   |-- <ArduinoJson-esphomelib> 5.13.3
|   |-- <ESP8266WiFi> 1.0
|   |-- <ESP8266mDNS>
|   |   |-- <ESP8266WiFi> 1.0
|   |-- <ESPAsyncTCP> 1.2.0
|   |   |-- <ESP8266WiFi> 1.0
|   |-- <ESP Async WebServer> 1.1.1
|   |   |-- <ESPAsyncTCP> 1.2.0
|   |   |   |-- <ESP8266WiFi> 1.0
|   |   |-- <ESP8266WiFi> 1.0
|   |   |-- <Hash> 1.0
|   |   |-- <ArduinoJson-esphomelib> 5.13.3
Compiling .pioenvs\esphome_test\src\main.cpp.o
Archiving .pioenvs\esphome_test\lib6d7\libESP8266WiFi.a
Archiving .pioenvs\esphome_test\libede\libESPAsyncTCP_ID305.a
Compiling .pioenvs\esphome_test\lib927\Hash\Hash.cpp.o
Compiling .pioenvs\esphome_test\lib927\Hash\sha1\sha1.c.o
Archiving .pioenvs\esphome_test\libd7a\libArduinoJson-esphomelib_ID3837.a
Compiling .pioenvs\esphome_test\libbb5\ESP Async WebServer_ID306\AsyncEventSource.cpp.o
Compiling .pioenvs\esphome_test\libbb5\ESP Async WebServer_ID306\AsyncWebSocket.cpp.o
Compiling .pioenvs\esphome_test\libbb5\ESP Async WebServer_ID306\SPIFFSEditor.cpp.o
Compiling .pioenvs\esphome_test\libbb5\ESP Async WebServer_ID306\WebAuthentication.cpp.o
Compiling .pioenvs\esphome_test\libbb5\ESP Async WebServer_ID306\WebHandlers.cpp.o
Compiling .pioenvs\esphome_test\libbb5\ESP Async WebServer_ID306\WebRequest.cpp.o
Compiling .pioenvs\esphome_test\libbb5\ESP Async WebServer_ID306\WebResponses.cpp.o
Compiling .pioenvs\esphome_test\libbb5\ESP Async WebServer_ID306\WebServer.cpp.o
Compiling .pioenvs\esphome_test\lib048\ESP8266mDNS\ESP8266mDNS.cpp.o
Compiling .pioenvs\esphome_test\liba97\ESPAsyncTCP_ID305@1.1.3\AsyncPrinter.cpp.o
src\main.cpp: In function 'void setup()':
src\main.cpp:34:93: error: no matching function for call to 'esphome::Application::make_global_variable(<brace-enclosed initializer list>)'
GlobalVariableComponent<int[4]> *gpio0_relays = App.make_global_variable<int[4]>({1,2,3,4});
^
src\main.cpp:34:93: note: candidates are:
In file included from .piolibdeps\esphome-core\src/esphome.h:1:0,
from src\main.cpp:3:
.piolibdeps\esphome-core\src/esphome/application.h:337:31: note: template<class T> esphome::GlobalVariableComponent<T>* esphome::Application::make_global_variable()
GlobalVariableComponent<T> *make_global_variable();
^
.piolibdeps\esphome-core\src/esphome/application.h:337:31: note:   template argument deduction/substitution failed:
src\main.cpp:34:93: note:   candidate expects 0 arguments, 1 provided
GlobalVariableComponent<int[4]> *gpio0_relays = App.make_global_variable<int[4]>({1,2,3,4});
^
In file included from .piolibdeps\esphome-core\src/esphome.h:1:0,
from src\main.cpp:3:
.piolibdeps\esphome-core\src/esphome/application.h:340:31: note: template<class T> esphome::GlobalVariableComponent<T>* esphome::Application::make_global_variable(T)
Archiving .pioenvs\esphome_test\lib927\libHash.aGlobalVariableComponent<T> *make_global_variable(T initial_value);

^
.piolibdeps\esphome-core\src/esphome/application.h:340:31: note:   template argument deduction/substitution failed:
src\main.cpp:34:93: note:   cannot convert '{1, 2, 3, 4}' (type '<brace-enclosed initializer list>') to type 'int*'
GlobalVariableComponent<int[4]> *gpio0_relays = App.make_global_variable<int[4]>({1,2,3,4});
^
src\main.cpp:35:94: error: no matching function for call to 'esphome::Application::make_global_variable(<brace-enclosed initializer list>)'
GlobalVariableComponent<int[4]> *gpio14_relays = App.make_global_variable<int[4]>({1,2,3,4});
^
src\main.cpp:35:94: note: candidates are:
In file included from .piolibdeps\esphome-core\src/esphome.h:1:0,
from src\main.cpp:3:
.piolibdeps\esphome-core\src/esphome/application.h:337:31: note: template<class T> esphome::GlobalVariableComponent<T>* esphome::Application::make_global_variable()
GlobalVariableComponent<T> *make_global_variable();
^
.piolibdeps\esphome-core\src/esphome/application.h:337:31: note:   template argument deduction/substitution failed:
src\main.cpp:35:94: note:   candidate expects 0 arguments, 1 provided
GlobalVariableComponent<int[4]> *gpio14_relays = App.make_global_variable<int[4]>({1,2,3,4});
^
In file included from .piolibdeps\esphome-core\src/esphome.h:1:0,
from src\main.cpp:3:
.piolibdeps\esphome-core\src/esphome/application.h:340:31: note: template<class T> esphome::GlobalVariableComponent<T>* esphome::Application::make_global_variable(T)
GlobalVariableComponent<T> *make_global_variable(T initial_value);
^
.piolibdeps\esphome-core\src/esphome/application.h:340:31: note:   template argument deduction/substitution failed:
src\main.cpp:35:94: note:   cannot convert '{1, 2, 3, 4}' (type '<brace-enclosed initializer list>') to type 'int*'
GlobalVariableComponent<int[4]> *gpio14_relays = App.make_global_variable<int[4]>({1,2,3,4});
^
Compiling .pioenvs\esphome_test\liba97\ESPAsyncTCP_ID305@1.1.3\ESPAsyncTCP.cpp.o
Compiling .pioenvs\esphome_test\liba97\ESPAsyncTCP_ID305@1.1.3\ESPAsyncTCPbuffer.cpp.o
*** [.pioenvs\esphome_test\src\main.cpp.o] Error 1
============================================== [ERROR] Took 3.74 seconds ==============================================
PS F:\Home Automation\@Projects\ESPHome> py -2 -m esphome test_switch.yaml version
Version: 1.11.2
```

**Additional information and things you've tried:**

https://community.home-assistant.io/t/using-and-initializing-array-variables-in-esphome/95991/3

https://discordapp.com/channels/429907082951524364/429907082955718657/555536353060716545

<!-- LEAVE THIS LINE AS-IS AND DON'T DELETE IT, OTHERWISE THE ISSUE WILL BE CLOSED AUTOMATICALLY. -->

--- a/platformio.ini
+++ b/platformio.ini
@@ -38,7 +38,7 @@ platform = espressif8266@1.8.0
 board = nodemcuv2
 framework = arduino
 lib_deps = ${common.lib_deps}
-build_flags = ${common.build_flags}
+build_flags = ${common.build_flags} -DESPHOME_LOG_LEVEL=6
 src_filter = ${common.src_filter} +<examples/livingroom8266/livingroom8266.cpp>
 
 [env:custombmp180]

--- a/src/esphome/api/api_server.cpp
+++ b/src/esphome/api/api_server.cpp
@@ -714,6 +714,12 @@ bool APIConnection::send_buffer(APIMessageType type) {
 }
 
 void APIConnection::loop() {
+  if (!network_is_connected()) {
+    // when network is disconnected force disconnect immediately
+    // don't wait for timeout
+    this->fatal_error_();
+    return;
+  }
   if (this->client_->disconnected()) {
     // failsafe for disconnect logic
     this->on_disconnect_();

--- a/src/esphome/api/api_server.cpp
+++ b/src/esphome/api/api_server.cpp
@@ -90,8 +90,10 @@ void APIServer::loop() {
         ESP_LOGE(TAG, "No client connected to API. Rebooting...");
         reboot("api");
       }
+      this->status_set_warning();
     } else {
       this->last_connected_ = now;
+      this->status_clear_warning();
     }
   }
 }

--- a/src/esphome/application.h
+++ b/src/esphome/application.h
@@ -334,7 +334,9 @@ class Application {
   template<typename T> GlobalVariableComponent<T> *make_global_variable();
 
   template<typename T> GlobalVariableComponent<T> *make_global_variable(T initial_value);
-  template<typename T> GlobalVariableComponent<T> *make_global_variable(std::array<typename std::remove_extent<T>::type, std::extent<T>::value> initial_value);
+  template<typename T>
+  GlobalVariableComponent<T> *make_global_variable(
+      std::array<typename std::remove_extent<T>::type, std::extent<T>::value> initial_value);
 
   /*           _    _ _______ ____  __  __       _______ _____ ____  _   _
    *      /\  | |  | |__   __/ __ \|  \/  |   /\|__   __|_   _/ __ \| \ | |
@@ -1235,7 +1237,8 @@ template<typename T> GlobalVariableComponent<T> *Application::make_global_variab
   return this->register_component(new GlobalVariableComponent<T>(initial_value));
 }
 template<typename T>
-GlobalVariableComponent<T> *Application::make_global_variable(std::array<typename std::remove_extent<T>::type, std::extent<T>::value> initial_value) {
+GlobalVariableComponent<T> *Application::make_global_variable(
+    std::array<typename std::remove_extent<T>::type, std::extent<T>::value> initial_value) {
   return this->register_component(new GlobalVariableComponent<T>(initial_value));
 }
 

--- a/src/esphome/application.h
+++ b/src/esphome/application.h
@@ -334,6 +334,7 @@ class Application {
   template<typename T> GlobalVariableComponent<T> *make_global_variable();
 
   template<typename T> GlobalVariableComponent<T> *make_global_variable(T initial_value);
+  template<typename T> GlobalVariableComponent<T> *make_global_variable(std::array<typename std::remove_extent<T>::type, std::extent<T>::value> initial_value);
 
   /*           _    _ _______ ____  __  __       _______ _____ ____  _   _
    *      /\  | |  | |__   __/ __ \|  \/  |   /\|__   __|_   _/ __ \| \ | |
@@ -1231,6 +1232,10 @@ template<typename T> GlobalVariableComponent<T> *Application::make_global_variab
 }
 
 template<typename T> GlobalVariableComponent<T> *Application::make_global_variable(T initial_value) {
+  return this->register_component(new GlobalVariableComponent<T>(initial_value));
+}
+template<typename T>
+GlobalVariableComponent<T> *Application::make_global_variable(std::array<typename std::remove_extent<T>::type, std::extent<T>::value> initial_value) {
   return this->register_component(new GlobalVariableComponent<T>(initial_value));
 }
 

--- a/src/esphome/automation.h
+++ b/src/esphome/automation.h
@@ -276,7 +276,8 @@ template<typename T> class GlobalVariableComponent : public Component {
  public:
   explicit GlobalVariableComponent();
   explicit GlobalVariableComponent(T initial_value);
-  explicit GlobalVariableComponent(std::array<typename std::remove_extent<T>::type, std::extent<T>::value> initial_value);
+  explicit GlobalVariableComponent(
+      std::array<typename std::remove_extent<T>::type, std::extent<T>::value> initial_value);
 
   T &value();
 

--- a/src/esphome/automation.h
+++ b/src/esphome/automation.h
@@ -276,6 +276,7 @@ template<typename T> class GlobalVariableComponent : public Component {
  public:
   explicit GlobalVariableComponent();
   explicit GlobalVariableComponent(T initial_value);
+  explicit GlobalVariableComponent(std::array<typename std::remove_extent<T>::type, std::extent<T>::value> initial_value);
 
   T &value();
 

--- a/src/esphome/automation.tcc
+++ b/src/esphome/automation.tcc
@@ -208,7 +208,8 @@ template<typename... Ts> ScriptStopAction<Ts...> *Script::make_stop_action() {
 template<typename T> GlobalVariableComponent<T>::GlobalVariableComponent() {}
 template<typename T> GlobalVariableComponent<T>::GlobalVariableComponent(T initial_value) : value_(initial_value) {}
 template<typename T>
-GlobalVariableComponent<T>::GlobalVariableComponent(std::array<typename std::remove_extent<T>::type, std::extent<T>::value> initial_value) {
+GlobalVariableComponent<T>::GlobalVariableComponent(
+    std::array<typename std::remove_extent<T>::type, std::extent<T>::value> initial_value) {
   memcpy(this->value_, initial_value.data(), sizeof(T));
 }
 template<typename T> T &GlobalVariableComponent<T>::value() { return this->value_; }

--- a/src/esphome/automation.tcc
+++ b/src/esphome/automation.tcc
@@ -207,6 +207,10 @@ template<typename... Ts> ScriptStopAction<Ts...> *Script::make_stop_action() {
 
 template<typename T> GlobalVariableComponent<T>::GlobalVariableComponent() {}
 template<typename T> GlobalVariableComponent<T>::GlobalVariableComponent(T initial_value) : value_(initial_value) {}
+template<typename T>
+GlobalVariableComponent<T>::GlobalVariableComponent(std::array<typename std::remove_extent<T>::type, std::extent<T>::value> initial_value) {
+  memcpy(this->value_, initial_value.data(), sizeof(T));
+}
 template<typename T> T &GlobalVariableComponent<T>::value() { return this->value_; }
 template<typename T> void GlobalVariableComponent<T>::setup() {
   if (this->restore_value_) {

--- a/src/esphome/binary_sensor/binary_sensor.cpp
+++ b/src/esphome/binary_sensor/binary_sensor.cpp
@@ -194,15 +194,15 @@ void MultiClickTrigger::on_state_(bool state) {
   MultiClickTriggerEvent evt = this->timing_[*this->at_index_];
 
   if (evt.max_length != 4294967294UL) {
-    ESP_LOGV(TAG, "A i=%u min=%u max=%u", *this->at_index_, evt.min_length, evt.max_length);
+    ESP_LOGV(TAG, "A i=%lu min=%u max=%u", *this->at_index_, evt.min_length, evt.max_length);
     this->schedule_is_valid_(evt.min_length);
     this->schedule_is_not_valid_(evt.max_length);
   } else if (*this->at_index_ + 1 != this->timing_.size()) {
-    ESP_LOGV(TAG, "B i=%u min=%u", *this->at_index_, evt.min_length);
+    ESP_LOGV(TAG, "B i=%lu min=%u", *this->at_index_, evt.min_length);
     this->cancel_timeout("is_not_valid");
     this->schedule_is_valid_(evt.min_length);
   } else {
-    ESP_LOGV(TAG, "C i=%u min=%u", *this->at_index_, evt.min_length);
+    ESP_LOGV(TAG, "C i=%lu min=%u", *this->at_index_, evt.min_length);
     this->is_valid_ = false;
     this->cancel_timeout("is_not_valid");
     this->set_timeout("trigger", evt.min_length, [this]() { this->trigger_(); });

--- a/src/esphome/binary_sensor/binary_sensor.cpp
+++ b/src/esphome/binary_sensor/binary_sensor.cpp
@@ -194,15 +194,15 @@ void MultiClickTrigger::on_state_(bool state) {
   MultiClickTriggerEvent evt = this->timing_[*this->at_index_];
 
   if (evt.max_length != 4294967294UL) {
-    ESP_LOGV(TAG, "A i=%lu min=%u max=%u", *this->at_index_, evt.min_length, evt.max_length);
+    ESP_LOGV(TAG, "A i=%u min=%u max=%u", *this->at_index_, evt.min_length, evt.max_length);  // NOLINT
     this->schedule_is_valid_(evt.min_length);
     this->schedule_is_not_valid_(evt.max_length);
   } else if (*this->at_index_ + 1 != this->timing_.size()) {
-    ESP_LOGV(TAG, "B i=%lu min=%u", *this->at_index_, evt.min_length);
+    ESP_LOGV(TAG, "B i=%u min=%u", *this->at_index_, evt.min_length);  // NOLINT
     this->cancel_timeout("is_not_valid");
     this->schedule_is_valid_(evt.min_length);
   } else {
-    ESP_LOGV(TAG, "C i=%lu min=%u", *this->at_index_, evt.min_length);
+    ESP_LOGV(TAG, "C i=%u min=%u", *this->at_index_, evt.min_length);  // NOLINT
     this->is_valid_ = false;
     this->cancel_timeout("is_not_valid");
     this->set_timeout("trigger", evt.min_length, [this]() { this->trigger_(); });

--- a/src/esphome/esppreferences.cpp
+++ b/src/esphome/esppreferences.cpp
@@ -80,7 +80,6 @@ static inline bool esp_rtc_user_mem_write(uint32_t index, uint32_t value) {
   auto *ptr = &ESP_RTC_USER_MEM[index];
 #ifdef USE_ESP8266_PREFERENCES_FLASH
   if (*ptr != value) {
-    ESP_LOGV(TAG, "RTC flash is dirty...");
     esp8266_preferences_modified = true;
   }
 #endif

--- a/src/esphome/esppreferences.cpp
+++ b/src/esphome/esppreferences.cpp
@@ -32,7 +32,7 @@ bool ESPPreferenceObject::load_() {
 
   bool valid = this->data_[this->length_words_] == this->calculate_crc_();
 
-  ESP_LOGVV(TAG, "LOAD %u: valid=%s, 0=0x%08X 1=0x%08X (Type=%u, CRC=0x%08X)", this->rtc_offset_, YESNO(valid),
+  ESP_LOGVV(TAG, "LOAD %zu: valid=%s, 0=0x%08X 1=0x%08X (Type=%u, CRC=0x%08X)", this->rtc_offset_, YESNO(valid),
             this->data_[0], this->data_[1], this->type_, this->calculate_crc_());
   return valid;
 }
@@ -45,7 +45,7 @@ bool ESPPreferenceObject::save_() {
   this->data_[this->length_words_] = this->calculate_crc_();
   if (!this->save_internal_())
     return false;
-  ESP_LOGVV(TAG, "SAVE %u: 0=0x%08X 1=0x%08X (Type=%u, CRC=0x%08X)", this->rtc_offset_, this->data_[0], this->data_[1],
+  ESP_LOGVV(TAG, "SAVE %zu: 0=0x%08X 1=0x%08X (Type=%u, CRC=0x%08X)", this->rtc_offset_, this->data_[0], this->data_[1],
             this->type_, this->calculate_crc_());
   return true;
 }

--- a/src/esphome/light/light_color_values.cpp
+++ b/src/esphome/light/light_color_values.cpp
@@ -82,18 +82,33 @@ void LightColorValues::normalize_color(const LightTraits &traits) {
     float max_value = fmaxf(this->get_red(), fmaxf(this->get_green(), this->get_blue()));
     if (traits.has_rgb_white_value()) {
       max_value = fmaxf(max_value, this->get_white());
-      this->set_white(this->get_white() / max_value);
+      if (max_value == 0.0f) {
+        this->set_white(1.0f);
+      } else {
+        this->set_white(this->get_white() / max_value);
+      }
     }
-    this->set_red(this->get_red() / max_value);
-    this->set_green(this->get_green() / max_value);
-    this->set_blue(this->get_blue() / max_value);
+    if (max_value == 0.0f) {
+      this->set_red(1.0f);
+      this->set_green(1.0f);
+      this->set_blue(1.0f);
+    } else {
+      this->set_red(this->get_red() / max_value);
+      this->set_green(this->get_green() / max_value);
+      this->set_blue(this->get_blue() / max_value);
+    }
   }
 
   if (traits.has_brightness() && this->get_brightness() == 0.0f) {
-    // 0% brightness means off
-    this->set_state(false);
-    // reset brightness to 100% (0% brightness is not allowed)
-    this->set_brightness(1.0f);
+    if (traits.has_rgb_white_value()) {
+      // 0% brightness for RGBW[W] means no RGB channel, but white channel on.
+      // do nothing
+    } else {
+      // 0% brightness means off
+      this->set_state(false);
+      // reset brightness to 100%
+      this->set_brightness(1.0f);
+    }
   }
 }
 

--- a/src/esphome/light/light_state.cpp
+++ b/src/esphome/light/light_state.cpp
@@ -438,7 +438,7 @@ void LightState::StateCall::perform() const {
     ESP_LOGD(TAG, "  Color Temperature: %.1f mireds", v.get_color_temperature());
   }
 
-  v.normalize_color(this->state_->output_->get_traits());
+  v.normalize_color(traits);
 
   if (traits.has_rgb() && (this->red_.has_value() || this->green_.has_value() || this->blue_.has_value())) {
     if (traits.has_rgb_white_value() && this->white_.has_value()) {

--- a/src/esphome/light/light_state.cpp
+++ b/src/esphome/light/light_state.cpp
@@ -124,14 +124,14 @@ void LightState::dump_json(JsonObject &root) {
 }
 
 struct LightStateRTCState {
-  bool state;
-  float brightness;
-  float red;
-  float green;
-  float blue;
-  float white;
-  float color_temp;
-  uint32_t effect;
+  bool state{false};
+  float brightness{1.0f};
+  float red{1.0f};
+  float green{1.0f};
+  float blue{1.0f};
+  float white{1.0f};
+  float color_temp{1.0f};
+  uint32_t effect{0};
 };
 
 void LightState::setup() {
@@ -143,9 +143,9 @@ void LightState::setup() {
   }
 
   this->rtc_ = global_preferences.make_preference<LightStateRTCState>(this->get_object_id_hash());
-  LightStateRTCState recovered;
-  if (!this->rtc_.load(&recovered))
-    return;
+  LightStateRTCState recovered{};
+  // Attempt to load from preferences, else fall back to default values from struct
+  this->rtc_.load(&recovered);
 
   auto call = this->make_call();
   call.set_state(recovered.state);

--- a/src/esphome/remote/remote_receiver.cpp
+++ b/src/esphome/remote/remote_receiver.cpp
@@ -326,8 +326,8 @@ void RemoteReceiverComponent::loop() {
     // TODO: Handle case when loop() is not called quickly enough to catch idle
     return;
 
-  ESP_LOGVV(TAG, "read_at=%u write_at=%u dist=%u now=%u end=%u", this->buffer_read_at_, write_at, dist, now,
-            this->buffer_[write_at]);
+  ESP_LOGVV(TAG, "read_at=%u write_at=%u dist=%u now=%u end=%u", s.buffer_read_at, write_at, dist, now,
+            s.buffer[write_at]);
 
   // Skip first value, it's from the previous idle level
   s.buffer_read_at = (s.buffer_read_at + 1) % s.buffer_size;
@@ -345,8 +345,8 @@ void RemoteReceiverComponent::loop() {
       break;
     }
 
-    ESP_LOGVV(TAG, "  i=%u buffer[%u]=%u - buffer[%u]=%u -> %d", i, this->buffer_read_at_,
-              this->buffer_[this->buffer_read_at_], prev, this->buffer_[prev], multiplier * delta);
+    ESP_LOGVV(TAG, "  i=%u buffer[%u]=%u - buffer[%u]=%u -> %d", i, s.buffer_read_at,
+              s.buffer[s.buffer_read_at], prev, s.buffer[prev], multiplier * delta);
     this->temp_.push_back(multiplier * delta);
     prev = s.buffer_read_at;
     s.buffer_read_at = (s.buffer_read_at + 1) % s.buffer_size;

--- a/src/esphome/remote/remote_receiver.cpp
+++ b/src/esphome/remote/remote_receiver.cpp
@@ -345,8 +345,8 @@ void RemoteReceiverComponent::loop() {
       break;
     }
 
-    ESP_LOGVV(TAG, "  i=%u buffer[%u]=%u - buffer[%u]=%u -> %d", i, s.buffer_read_at,
-              s.buffer[s.buffer_read_at], prev, s.buffer[prev], multiplier * delta);
+    ESP_LOGVV(TAG, "  i=%u buffer[%u]=%u - buffer[%u]=%u -> %d", i, s.buffer_read_at, s.buffer[s.buffer_read_at], prev,
+              s.buffer[prev], multiplier * delta);
     this->temp_.push_back(multiplier * delta);
     prev = s.buffer_read_at;
     s.buffer_read_at = (s.buffer_read_at + 1) % s.buffer_size;

--- a/src/esphome/sensor/cse7766.cpp
+++ b/src/esphome/sensor/cse7766.cpp
@@ -43,34 +43,8 @@ bool CSE7766Component::check_byte_() {
   uint8_t index = this->raw_data_index_;
   uint8_t byte = this->raw_data_[index];
   if (index == 0) {
-    // Header - 0x55
-    // These messages are verbose because the CSE7766
-    // reports these when no load is attached. If the checksum doesn't match
-    // though, then we should print a warning.
-    if (byte == 0xAA) {
-      ESP_LOGV(TAG, "CSE7766 not calibrated!");
-      return false;
-    }
-    if ((byte & 0xF0) == 0xF0) {
-      ESP_LOGV(TAG, "CSE7766 reports abnormal hardware: (0x%02X)", byte);
-      if ((byte >> 3) & 1) {
-        ESP_LOGV(TAG, "  Voltage cycle exceeds range.");
-      }
-      if ((byte >> 2) & 1) {
-        ESP_LOGV(TAG, "  Current cycle exceeds range.");
-      }
-      if ((byte >> 1) & 1) {
-        ESP_LOGV(TAG, "  Power cycle exceeds range.");
-      }
-      if ((byte >> 0) & 1) {
-        ESP_LOGV(TAG, "  Coefficient storage area is abnormal.");
-      }
-      return false;
-    }
-    if (byte != 0x55) {
-      ESP_LOGV(TAG, "Invalid Header Start from CSE7766: 0x%02X!", byte);
-      return false;
-    }
+    // Header, usually 0x55, contains data about calibration etc.
+    // this is validated in parse_data_
     return true;
   }
 
@@ -103,6 +77,18 @@ void CSE7766Component::parse_data_() {
     ESP_LOGVV(TAG, "  i=%u: 0b" BYTE_TO_BINARY_PATTERN " (0x%02X)", i, BYTE_TO_BINARY(this->raw_data_[i]),
               this->raw_data_[i]);
   }
+
+  uint8_t header1 = this->raw_data_[0];
+  if (header1 == 0xAA) {
+    ESP_LOGW(TAG, "CSE7766 not calibrated!");
+    return;
+  }
+  if ((header1 & 0xF0) == 0xF0 && ((header1 >> 0) & 1) == 1) {
+    ESP_LOGW(TAG, "CSE7766 reports abnormal hardware: (0x%02X)", header1);
+    ESP_LOGW(TAG, "  Coefficient storage area is abnormal.");
+    return;
+  }
+
   const uint32_t now = micros();
   const float d = (now - this->last_reading_) / 1000.0f;
   this->last_reading_ = now;
@@ -116,19 +102,25 @@ void CSE7766Component::parse_data_() {
 
   uint8_t adj = this->raw_data_[20];
 
-  if ((adj >> 6 != 0) && voltage_cycle != 0) {
+  if ((adj >> 6 != 0) && voltage_cycle != 0 &&
+      // voltage cycle exceeds range
+      ((header1 >> 3) & 1) == 0) {
     // voltage cycle of serial port outputted is a complete cycle;
     float voltage = voltage_calib / float(voltage_cycle);
     this->voltage_acc_ += voltage * d;
   }
 
-  if ((adj >> 5 != 0) && power_cycle != 0 && current_cycle != 0) {
+  if ((adj >> 5 != 0) && power_cycle != 0 && current_cycle != 0 &&
+      // current cycle exceeds range
+      ((header1 >> 2) & 1) == 0) {
     // indicates current cycle of serial port outputted is a complete cycle;
     float current = current_calib / float(current_cycle);
     this->current_acc_ += current * d;
   }
 
-  if ((adj >> 4 != 0) && power_cycle != 0) {
+  if ((adj >> 4 != 0) && power_cycle != 0 &&
+      // power cycle exceeds range
+      ((header1 >> 1) & 1) == 0) {
     // power cycle of serial port outputted is a complete cycle;
     float active_power = power_calib / float(power_cycle);
     this->power_acc_ += active_power * d;

--- a/src/esphome/sensor/rotary_encoder.cpp
+++ b/src/esphome/sensor/rotary_encoder.cpp
@@ -88,7 +88,7 @@ void ICACHE_RAM_ATTR HOT RotaryEncoderSensorStore::gpio_intr(RotaryEncoderSensor
   if (arg->pin_b->digital_read())
     input_state |= STATE_PIN_B_HIGH;
 
-  uint8_t new_state = STATE_LOOKUP_TABLE[input_state];
+  uint16_t new_state = STATE_LOOKUP_TABLE[input_state];
   if ((new_state & arg->resolution & STATE_HAS_INCREMENTED) != 0) {
     if (arg->counter < arg->max_value)
       arg->counter++;

--- a/src/esphome/wifi_component.cpp
+++ b/src/esphome/wifi_component.cpp
@@ -174,7 +174,7 @@ void WiFiComponent::start_connecting(const WiFiAP &ap, bool two) {
   } else {
     ESP_LOGV(TAG, "  BSSID: Not Set");
   }
-  ESP_LOGV(TAG, "  Password: " LOG_SECRET("'%s'"), ap.get_password());
+  ESP_LOGV(TAG, "  Password: " LOG_SECRET("'%s'"), ap.get_password().c_str());
   if (ap.get_channel().has_value()) {
     ESP_LOGV(TAG, "  Channel: %u", *ap.get_channel());
   } else {
@@ -779,6 +779,12 @@ const char *get_disconnect_reason_str(uint8_t reason) {
 }
 
 void WiFiComponent::wifi_event_callback(System_Event_t *event) {
+#ifdef ESPHOME_LOG_HAS_VERBOSE
+  // TODO: this callback is called while in cont context, so delay will fail
+  // We need to defer the log messages until we're out of this context
+  // only affects verbose log level
+  // reproducible by enabling verbose log level and letting the ESP disconnect and
+  // then reconnect to WiFi.
   switch (event->event) {
     case EVENT_STAMODE_CONNECTED: {
       auto it = event->event_info.connected;
@@ -794,7 +800,7 @@ void WiFiComponent::wifi_event_callback(System_Event_t *event) {
       char buf[33];
       memcpy(buf, it.ssid, it.ssid_len);
       buf[it.ssid_len] = '\0';
-      ESP_LOGW(TAG, "Event: Disconnected ssid='%s' bssid=%s reason='%s'", buf, format_mac_addr(it.bssid).c_str(),
+      ESP_LOGV(TAG, "Event: Disconnected ssid='%s' bssid=%s reason='%s'", buf, format_mac_addr(it.bssid).c_str(),
                get_disconnect_reason_str(it.reason));
       break;
     }
@@ -844,6 +850,7 @@ void WiFiComponent::wifi_event_callback(System_Event_t *event) {
     default:
       break;
   }
+#endif
 
   if (event->event == EVENT_STAMODE_DISCONNECTED) {
     global_wifi_component->error_from_callback_ = true;

